File: workers/fpm/worker.py
import json
import subprocess
import sys
from pathlib import Path


FPM = Path(r"C:\Users\matth\AppData\Local\fortran-lang\fpm\fpm.exe")
if not FPM.exists():
    FPM = Path("fpm")


def send(message):
    sys.stdout.write(json.dumps(message) + "\n")
    sys.stdout.flush()


for line in sys.stdin:
    message = json.loads(line)
    if message["type"] == "HELLO":
        send(
            {
                "type": "WELCOME",
                "protocol": "ofp/1",
                "workerId": "fpm-plan-01",
                "language": "fpm",
                "runtimeVersion": "0.13.0",
                "workerVersion": "0.1.0",
                "status": "ready",
            }
        )
        send(
            {
                "type": "REGISTER",
                "protocol": "ofp/1",
                "workerId": "fpm-plan-01",
                "language": "fpm",
                "runtimeVersion": "0.13.0",
                "workerVersion": "0.1.0",
                "capabilities": [{"name": "dev.fortran-tooling"}],
            }
        )
    elif message["type"] == "REGISTER_ACK":
        continue
    elif message["type"] == "JOB_START":
        if message["capability"] != "dev.fortran-tooling":
            send({"type": "JOB_ERROR", "jobId": "unknown", "error": "unsupported capability"})
            break
        send({"type": "JOB_ACCEPTED", "jobId": message["jobId"], "status": "running"})
        send({"type": "JOB_LOG", "jobId": message["jobId"], "severity": "info", "message": "starting dev.fortran-tooling"})
        completed = subprocess.run([str(FPM), "--version"], capture_output=True, text=True, check=False)
        version = (completed.stdout or completed.stderr or "").strip().splitlines()
        version_text = version[0] if version else "unknown"
        package = str(message.get("input", {}).get("package", "") or "")
        send(
            {
                "type": "JOB_RESULT",
                "jobId": message["jobId"],
                "output": {
                    "tool": "fpm",
                    "version": version_text,
                    "requestedPackage": package,
                    "suggestedCommand": "fpm new demo" if not package else f"fpm build --profile {package}",
                },
            }
        )
        break
    elif message["type"] == "JOB_CANCEL":
        send({"type": "JOB_CANCELLED", "jobId": message.get("jobId", "job-unknown"), "status": "cancelled"})
        break
    elif message["type"] == "SHUTDOWN":
        send({"type": "SHUTDOWN_ACK", "workerId": "fpm-plan-01", "status": "stopped"})
        break

File: workers/ruby/worker.rb
require "json"

REGISTER = {
  type: "REGISTER",
  protocol: "ofp/1",
  workerId: "ruby-rules-01",
  language: "ruby",
  runtimeVersion: RUBY_VERSION,
  workerVersion: "0.1.0",
  capabilities: [{ name: "rules.evaluate" }]
}

WELCOME = {
  type: "WELCOME",
  protocol: "ofp/1",
  workerId: "ruby-rules-01",
  language: "ruby",
  runtimeVersion: RUBY_VERSION,
  workerVersion: "0.1.0",
  status: "ready"
}

def send_message(message)
  STDOUT.puts(JSON.generate(message))
  STDOUT.flush
end

ARGF.each_line do |line|
  message = JSON.parse(line)
  case message["type"]
  when "HELLO"
    send_message(WELCOME)
    send_message(REGISTER)
  when "REGISTER_ACK"
    next
  when "JOB_START"
    if message["capability"] != "rules.evaluate"
      send_message(type: "JOB_ERROR", jobId: message["jobId"], error: "unsupported capability")
      next
    end

    send_message(type: "JOB_ACCEPTED", jobId: message["jobId"], status: "running")
    send_message(type: "JOB_LOG", jobId: message["jobId"], severity: "info", message: "starting rules.evaluate")

    rows = message["input"]["previous"]["rows"]
    classified = rows.map do |row|
      risk = if !row["active"]
               "inactive"
             elsif row["amount"] >= 1000
               "priority"
             elsif row["amount"] >= 500
               "standard"
             else
               "low"
             end
      row.merge("risk" => risk)
    end

    send_message(type: "JOB_PROGRESS", jobId: message["jobId"], progress: 0.5, metadata: { stage: "classification" })
    send_message(
      type: "JOB_RESULT",
      jobId: message["jobId"],
      output: {
        rows: classified,
        counts: classified.group_by { |row| row["risk"] }.transform_values(&:length)
      }
    )
  when "SHUTDOWN"
    send_message(type: "SHUTDOWN_ACK", workerId: "ruby-rules-01", status: "stopped")
    break
  end
end
